pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/DataTypeTransformer.java

```diff
@@ -92,6 +92,13 @@ public GenericRow transform(GenericRow record) {
           // Single-value column
           source = PinotDataType.getSingleValueType(value.getClass());
         }
+
+        if (source == PinotDataType.BYTES && dest == PinotDataType.STRING && value instanceof byte[]) {
```

Comment:
This will introduce more severe backward compatibility issue, where the bytes values are expected to be converted using hex encoding. Let's not try to handle this

Reply:
Without the provided solution, a Fixed type cannot be converted to byte[]. Currently if the input is of type Fixed and the corresponding pinot column is of type BYTES, we get the following exception in `DataTypeTransformer`:
```
java.lang.RuntimeException: Caught exception while transforming data type for column: bytesColumn
	at org.apache.pinot.segment.local.recordtransformer.DataTypeTransformer.transform(DataTypeTransformer.java:118)
	at org.apache.pinot.segment.local.recordtransformer.CompositeTransformer.transform(CompositeTransformer.java:90)
	at org.apache.pinot.segment.local.segment.creator.TransformPipeline.processPlainRow(TransformPipeline.java:97)
	at org.apache.pinot.segment.local.segment.creator.TransformPipeline.processRow(TransformPipeline.java:92)
	at org.apache.pinot.core.data.manager.realtime.LLRealtimeSegmentDataManager.processStreamEvents(LLRealtimeSegmentDataManager.java:554)
	at org.apache.pinot.core.data.manager.realtime.LLRealtimeSegmentDataManager.consumeLoop(LLRealtimeSegmentDataManager.java:429)
	at org.apache.pinot.core.data.manager.realtime.LLRealtimeSegmentDataManager$PartitionConsumer.run(LLRealtimeSegmentDataManager.java:622)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalArgumentException: Value: [0, 1, 2, 3] is not Hex encoded
	at org.apache.pinot.spi.utils.BytesUtils.toBytes(BytesUtils.java:39)
	at org.apache.pinot.common.utils.PinotDataType$11.toBytes(PinotDataType.java:646)
	at org.apache.pinot.common.utils.PinotDataType$13.convert(PinotDataType.java:762)
	at org.apache.pinot.segment.local.recordtransformer.DataTypeTransformer.transform(DataTypeTransformer.java:112)
	... 7 more
Caused by: org.apache.commons.codec.DecoderException: Illegal hexadecimal character [ at index 0
	at org.apache.commons.codec.binary.Hex.toDigit(Hex.java:369)
	at org.apache.commons.codec.binary.Hex.decodeHex(Hex.java:107)
	at org.apache.commons.codec.binary.Hex.decodeHex(Hex.java:77)
	at org.apache.pinot.spi.utils.BytesUtils.toBytes(BytesUtils.java:37)
	... 10 more
java.lang.RuntimeException: Caught exception while transforming data type for column: bytesColumn
	at org.apache.pinot.segment.local.recordtransformer.DataTypeTransformer.transform(DataTypeTransformer.java:118)
	at org.apache.pinot.segment.local.recordtransformer.CompositeTransformer.transform(CompositeTransformer.java:90)
	at org.apache.pinot.segment.local.segment.creator.TransformPipeline.processPlainRow(TransformPipeline.java:97)
	at org.apache.pinot.segment.local.segment.creator.TransformPipeline.processRow(TransformPipeline.java:92)
	at org.apache.pinot.core.data.manager.realtime.LLRealtimeSegmentDataManager.processStreamEvents(LLRealtimeSegmentDataManager.java:554)
	at org.apache.pinot.core.data.manager.realtime.LLRealtimeSegmentDataManager.consumeLoop(LLRealtimeSegmentDataManager.java:429)
	at org.apache.pinot.core.data.manager.realtime.LLRealtimeSegmentDataManager$PartitionConsumer.run(LLRealtimeSegmentDataManager.java:622)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalArgumentException: Value: [0, 1, 2, 3] is not Hex encoded
	at org.apache.pinot.spi.utils.BytesUtils.toBytes(BytesUtils.java:39)
	at org.apache.pinot.common.utils.PinotDataType$11.toBytes(PinotDataType.java:646)
	at org.apache.pinot.common.utils.PinotDataType$13.convert(PinotDataType.java:762)
	at org.apache.pinot.segment.local.recordtransformer.DataTypeTransformer.transform(DataTypeTransformer.java:112)
	... 7 more
Caused by: org.apache.commons.codec.DecoderException: Illegal hexadecimal character [ at index 0
	at org.apache.commons.codec.binary.Hex.toDigit(Hex.java:369)
	at org.apache.commons.codec.binary.Hex.decodeHex(Hex.java:107)
	at org.apache.commons.codec.binary.Hex.decodeHex(Hex.java:77)
	at org.apache.pinot.spi.utils.BytesUtils.toBytes(BytesUtils.java:37)
	... 10 more

Reply:
In other words, the only way to use Fixed type in the input data is to have corresponding String column in pinot schema. With the provided bug fix, that behavior doesn't change for existing tables. For new tables, the correct way is to have pinot column as BYTES not STRING.

Reply:
I don't follow. With the fix, the Fixed type should be directly read as byte[] and no conversion should be needed

Reply:
@Jackie-Jiang I briefly met with Sajjad and he mentioned that this type change code won't be needed for new tables. But, he wants to keep the backward compatibility for the existing tables (where the input data is Avro's Fixed type column and Pinot is configured with `string`column ) since he cannot afford the backfill right away.

@sajjad-moradi Can you provide more context to Jackie?

Reply:
I looked into Avro types again and realized that byte array can be represented by "FIXED" types as well as "BYTES" type. We have more complicated situation now. 
Let's assume the value for the input avro field is 'new byte[] {0, 1, 2, 3}'. We have the following four scenarios for different combinations of avro field type and pinot column type:
![image](https://user-images.githubusercontent.com/8548220/201789886-b11fb0c3-3787-4949-97b3-4cc8b82fde6c.png)
No matter if we keep or remove the backward-compatibility added in this this PR, we'll have backward incompatiblity.

Reply:
cc: @gviedma @mcvsubbu

Reply:
The only backward incompatibility is when user expects to store bytes as string [123, 45, 6], which IMO should never happen, so I won't worry too much about it. We don't want to specially handle it because it can break the case when user expects to strong bytes as hex encoded string, which is the way we support bytes before

Reply:
Simplified the PR to just go with the fix.